src/CorePolicy.cc

```diff
@@ -0,0 +1,201 @@
+
+using PerfUtils::Cycles;
+
+/**
+  * Arachne will attempt to increase the number of cores if the idle core
+  * fraction (computed as idlePercentage * numSharedCores) is less than this
+  * number.
+  */
+double maxIdleCoreFraction = 0.1;
+
+/**
+  * Arachne will attempt to increase the number of cores if the load factor
+  * increase beyond this threshold.
+  */
+double loadFactorThreshold = 1.0;
+
+/**
+  * Only here to satisfy compiling benchmark.
+  */
+double maxUtilization = 0.9;
+
+/**
+  * Save the core fraction at which we ramped up based on load factor, so we
+  * can decide whether to ramp down.
+  */
+double *utilizationThresholds;
+
+/**
+  * The difference in load, expressed as a fraction of a core, between a
+  * ramp-down threshold and the corresponding ramp-up threshold.
+  */
+double idleCoreFractionHysteresis = 0.2;
+
+/**
+  * The percentage of slots which are occupied that will prevent reducing the
+  * number of cores.
+  */
+double SLOT_OCCUPANCY_THRESHOLD = 0.5;
+
+/**
+  * The period in ns over which we measure before deciding to reduce the number
+  * of cores we use.
+  */
+const uint64_t MEASUREMENT_PERIOD = 50 * 1000 * 1000;
+
+void coreLoadEstimator();
+
+/**
+  * Bootstrap the core load estimator thread.  This must be done separately
+  * from the CorePolicy constructor because the constructor must run before
+  * Arachne adds any cores, but the core load estimator must run after
+  * Arachne has cores.
+  */
+void CorePolicy::bootstrapLoadEstimator(bool disableLoadEstimation) {
+    if (!disableLoadEstimation) {
+        utilizationThresholds = new double[Arachne::maxNumCores];
+        Arachne::createThread(baseClass, coreLoadEstimator);
+    }
+}
+
+/**
+  * Choose a core to deschedule.  Return its coreId.
+  */
+int CorePolicy::chooseRemovableCore() {
+    threadCoreMapEntry* entry = threadCoreMap[baseClass];
+    uint8_t minLoaded =
+        Arachne::occupiedAndCount[entry->map[0]]->load().numOccupied;
+    int minCoreId = entry->map[0];
+    for (uint32_t i = 1; i < threadCoreMap[baseClass]->numFilled; i++) {
+        uint32_t coreId = entry->map[i];
+        if (Arachne::occupiedAndCount[coreId]->load().numOccupied < minLoaded) {
+            minLoaded = Arachne::occupiedAndCount[coreId]->load().numOccupied;
+            minCoreId = coreId;
+        }
+    }
+    return minCoreId;
+}
+
+/**
+  * Update threadCoreMap when a new core is added.  Takes in the coreId
+  * of the new core.
+**/
+void CorePolicy::addCore(int coreId) {
+  threadCoreMapEntry* entry = threadCoreMap[baseClass];
+  entry->map[entry->numFilled] = coreId;
+  entry->numFilled++;
+}
+
+/**
+  * Update threadCoreMap when a core is removed.  Takes in the coreId
+  * of the doomed core.
+**/
+void CorePolicy::removeCore(int coreId) {
+  threadCoreMapEntry* entry = threadCoreMap[baseClass];
+  entry->numFilled--;
+  int saveCoreId = entry->map[entry->numFilled];
+  for (uint32_t i = 0; i < entry->numFilled; i++) {

```

Comment:
Are there concurrency issues if new threads are created while threadCoreMap is being updated? Let's discuss. If nothing else, concurrency needs to be mentioned in the documentation for this variable (I suspect you have followed certain constraints in your coding, but I don't see any place where those constraints are documented).